### PR TITLE
Upgrade

### LINKED ISSUE #38: compatibility branch for vue-apollo 3.x
Need to work on a compatibility version for vue-apollo 3.x and apollo-client 2.x

@Akryum it would be great to catch up on this issue together. Thanks!

### PR DESCRIPTION
**Pull request by Hendrixer**
✨  Update to `vue-apollo 3.x` and `apollo 2.x`. closes #38 

**Breaking changes**
* No need for `networkInterfaces` anymore. You **must** export config Objects for `ApolloClient`.
  * Allows users to compose their `Links` and still have multiple clients.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,32 +40,31 @@ Example (`nuxt.config.js`):
...
-  // apollo-client options
-  // return networkInterface
-
-  // alternative return a object with constructor options of apollo-client
+  // return the an object with additional apollo-client options
   return {
-    networkInterface,
+    link: appLink,
+    cache: new InMemoryCache(),
     dataIdFromObject: o => o.id
```
Comment (Akryum) [inline README.md]: `dataIdFromObject` is now a cache option. Also, it should really be:

```js
dataIdFromObject: object => `${object.__typename}:${object.id}`
```

File: index.js [hunk 2]
```diff
@@ -3,28 +3,28 @@ const path = require('path')
...
   this.addPlugin({
-    src: path.join(__dirname, 'plugin.js'),
-    options: options
+    options: options,
+    src: path.join(__dirname, 'plugin.js')
   })
 
   // Add vue-apollo and apollo-client in common bundle
-  this.addVendor(['vue-apollo', 'apollo-client'])
+  this.addVendor(['vue-apollo', 'apollo-client', 'apollo-cache-inmemory'])
```
Comment (Akryum) [inline index.js]: What about those?

```
apollo-link
apollo-link-http
apollo-link-batch-http
apollo-link-retry
apollo-link-context
apollo-link-error
apollo-link-ws
apollo-utilities
```

File: plugin.js [hunk 3]
```diff
@@ -13,19 +14,22 @@ export default (ctx) => {
...
-    networkInterface = networkInterface.default(ctx) || networkInterface(ctx)
+  <% Object.keys(options.clientConfigs).forEach((key) => { %>
+    let client = require('<%= options.clientConfigs[key] %>')
+    // es6 module default export or not
+    client = client.default(ctx) || client(ctx)
 
     const opts = isServer ? {
-        ssrMode: true
+        ssrMode: true,
+        cache: new InMemoryCache()
```
Comment (Akryum) [inline plugin.js]: Why creating a new cache and not getting the one from the config?

File: plugin.js [hunk 4]
```diff
@@ -13,19 +14,22 @@ export default (ctx) => {
...
     } : {
-      initialState: window.__NUXT__ ? window.__NUXT__.apollo.<%= key === 'default' ? 'defaultClient' : key %> : null,
+      initialState: window.__NUXT__ ? window.__NUXT__.apollo.<%= key === 'default' ? 'defaultLink' : key %> : null,
```
Comment (Akryum) [inline plugin.js]: There is no `intialState` anymore. See [the docs](https://github.com/Akryum/vue-apollo/tree/next#creating-the-apollo-clients).

File: plugin.js [hunk 5]
```diff
@@ -13,19 +14,22 @@ export default (ctx) => {
...
       ssrForceFetchDelay: 100,
+      cache: new InMemoryCache(),
```
Comment (Akryum) [inline plugin.js]: See comments above.

File: index.js (hunk 2 above)
Comment (Hendrixer) [reply index.js]: I've actually never really used nuxt until a week ago, but I was under the impression that these are modules that this nuxt module is using and needs to bundle, correct? This nuxt module does not make `Links`, the user would supply those, which would then get bundled.

File: plugin.js (hunk 3 above)
Comment (Hendrixer) [reply plugin.js]: fasho

File: plugin.js (hunk 4 above)
Comment (Hendrixer) [reply plugin.js]: dope

File: plugin.js (hunk 5 above)
Comment (Hendrixer) [reply plugin.js]: will do, thanks

File: index.js (hunk 2 above)
Comment (Akryum) [reply index.js]: Yes I think you are right.

File: plugin.js (hunk 3 above)
Comment (Akryum) [reply plugin.js]: You are still unnecessary creating a cache when one is already provided.

File: index.js [hunk 6]
```diff
@@ -3,28 +3,28 @@ const path = require('path')
...
-  options.networkInterfaces = options.networkInterfaces || {}
+  options.clientConfigs = options.clientConfigs || {}
 
   // Check network interfaces valid definition
-  const networkInterfaces = options.networkInterfaces
-  if (Object.keys(networkInterfaces).length === 0) throw new Error('[Apollo module] No network interfaces found in apollo configuration')
-  if (!networkInterfaces.default) throw new Error('[Apollo module] No default network interface found in apollo configuration')
+  const clientConfigs = options.clientConfigs
+  if (Object.keys(clientConfigs).length === 0) throw new Error('[Apollo module] No clientConfigs found in apollo configuration')
+  if (!clientConfigs.default) throw new Error('[Apollo module] No default link found in apollo configuration')
```
Comment (Akryum) [inline index.js]: 'client configuration' instead of 'link'

File: index.js [hunk 7]
```diff
@@ -3,28 +3,28 @@ const path = require('path')
...
 
-  // Sanitize networkInterfaces option
-  Object.keys(networkInterfaces).forEach((key) => {
-    if (typeof networkInterfaces[key] !== 'string' || (typeof networkInterfaces[key] === 'string' && /^https?:\/\//.test(networkInterfaces[key]))) {
-      throw new Error(`[Apollo module] Network interface "${key}" should be a path to a network interface.`)
+  // Sanitize clientConfigs option
+  Object.keys(clientConfigs).forEach((key) => {
+    if (typeof clientConfigs[key] !== 'string' || (typeof clientConfigs[key] === 'string' && /^https?:\/\//.test(clientConfigs[key]))) {
+      throw new Error(`[Apollo module] Network interface "${key}" should be a path to an exported Apollo Client config object.`)
```
Comment (Akryum) [inline index.js]: It's not a 'Network interface' anymore. :smile:

File: plugin.js (hunk 3 above)
Comment (Akryum) [reply plugin.js]: The code should look like this:

```js
const cache = client.cache || new InMemoryCache()

const opts = isServer ? {
  ssrMode: true,
} : {
  ssrForceFetchDelay: 100,
  connectToDevTools: isDev
}

if (!isServer) {
  cache.restore(window.__NUXT__ ? window.__NUXT__.apollo.<%= key === 'default' ? 'defaultClient' : key %> : null)
}

const finalOptions = Object.assign({}, opts, client, { cache })
```

**Comment by Akryum** [comment]
Thanks for the PR @Hendrixer! I suggested some more changes. Keep up the good work! :+1:

File: plugin.js (hunk 3 above)
Comment (Hendrixer) [reply plugin.js]: ok, yea, using the same singleton, I get what you're saying.

**Comment by gdi2290** [comment]
looking forward to seeing this merged!

**Comment by dohomi** [comment]
@Hendrixer thanks your PR works great. @Atinux can you double check on this and publish a new version in case you confirm the changes? Thanks

**Comment by dohomi** [comment]
@bjunc can you check if this resolves your opened PR as well?

**Comment by Atinux** [comment]
@dohomi what is your NPM username? So I can add you as collaborator to publish a new major release for it? You are handling this module perfectly and I trust you to improve and maintain it :)

**Comment by Akryum** [comment]
@Hendrixer Looking good! Nice job! 👍

**Comment by bjunc** [comment]
Couple comments:

* @dohomi No unfortunately, this has the same problem of redeclaring `client` as the current logic (line 18).  Using EJS to prefix the key is what I did in my pull request, which would also apply here.
* In the readme, recommend naming the instantiated link `httpLink` instead of `appLink` (per Apollo naming convention)
* Also, due to the common practice of assigning a token to the interface/link, I think it'd be helpful to show a simple example of how middleware is instantiated, creates a new link, which is then returned.  Quick example:

```js
import { ApolloLink } from 'apollo-link'
import { HttpLink } from 'apollo-link-http'
import { InMemoryCache } from 'apollo-cache-inmemory'

export default (ctx) => {
  const httpLink = new HttpLink({ uri: 'http://localhost:8000/graphql' })

  // auth token
  let token = ctx.isServer ? ctx.req.session : window.__NUXT__.state.session

  // middleware
  const middlewareLink = new ApolloLink((operation, forward) => {
    operation.setContext({
      headers: { authorization: `Bearer ${token}` }
    })
    return forward(operation)
  })
  const link = middlewareLink.concat(httpLink)
  return { link, cache: new InMemoryCache() }
}
```

**Comment by Hendrixer** [comment]
@bjunc 

* I like the example, I'll update
* also, folder names in README makes sense
* I don't think this PR should fix the issue you were describing because it's solving a different one. Just upgrading. After this is merged, I suggest updating you PR.

**Comment by dohomi** [comment]
@Atinux unfortunately my npm name is `dohomi`, thanks for adding me there (but I have to admit that I never maintained a npm package before)

I agree to the changes in the README @bjunc suggested for me this PR looks ready to merge

**Comment by Hendrixer** [comment]
@dohomi its easy! Just keep up with semver and release some tags, this is a tiny package

**Comment by Atinux** [comment]
@dohomi I added you as collaborator for https://www.npmjs.com/package/@nuxtjs/apollo

I suggest your to use https://github.com/sindresorhus/np to make a new release, for this one, since it makes a breaking change, it has to be a `major`: `np major`

Then you can use https://github.com/zeit/release to make a draft of the release based on the commits on GitHub :)

Don't worry it's a simple module, welcome to the community!

Also, thanks @Hendrixer for the PR!

**Pull request merged**